### PR TITLE
Exploring the sensitivity of analog inference, analog-aware retraining and in-analog training to the non-idealities of Analog Crossbars

### PR DESCRIPTION
**Pull request by AbinandN**
## Description

A notebook that explores and experiments with some of the non-idealities of analog crossbar implementations.

### CODE REVIEW

File: notebooks/examples/analog_sensitivity_LeNet5.ipynb [hunk 1]
```diff
@@ -0,0 +1,1483 @@
...
+   "id": "14ceee0f",
+   "metadata": {
+    "slideshow": {
+     "slide_type": "slide"
+    }
+   },
+   "source": [
+    "## Experiment 1: Analog Inference\n",
+    "\n",
+    "As a preliminary experiment, we perform \"analog\" inference of the previously trained digital LeNet with a floating point synaptic device model, while ignoring the non-idealities of peripheral circuits. "
```
Comment (maljoras) [inline notebooks/examples/analog_sensitivity_LeNet5.ipynb:522]: This is indeed just the floating point model if you use the FloatingPointRPUConfig. Could you maybe add a sentence say this? Something like: Using FloatingPointRPUConfig is a way to run the floating point reference model using the "analog" layers and should result in the same accuracy as the model with native pytorch layers.

File: notebooks/examples/analog_sensitivity_LeNet5.ipynb [hunk 2]
```diff
@@ -0,0 +1,1483 @@
...
+   "id": "fdf3ea57",
+   "metadata": {
+    "slideshow": {
+     "slide_type": "slide"
+    }
+   },
+   "source": [
+    "## Experiment 2: Parameters of Analog Inference\n",
+    "\n",
+    "In this experiment, we introduce a few of the non-idealities involved in analog inference. Experimenting with different settings for these factors shows the clear impact of analog inference on accuracy."
```
Comment (maljoras) [inline notebooks/examples/analog_sensitivity_LeNet5.ipynb:598]: You could add that for analog inference, only non-idealities in the forward pass will be affecting the accuracy, as backward (and update) is considered perfect and ideal. We should always use InferenceRPUConfig for analog inference chips.

File: notebooks/examples/analog_sensitivity_LeNet5.ipynb [hunk 3]
```diff
@@ -0,0 +1,1483 @@
...
+   "outputs": [
+    {
+     "data": {
+      "application/vnd.jupyter.widget-view+json": {
+       "model_id": "6769a08682f1416e83cd2eb7b209f02f",
+       "version_major": 2,
+       "version_minor": 0
+      },
+      "text/plain": [
+       "Dropdown(description='Device:', options=('FloatingPointDevice', 'ConstantStepDevice', 'LinearStepDevice', 'Ide…"
```
Comment (maljoras) [inline notebooks/examples/analog_sensitivity_LeNet5.ipynb:620]: It makes no sense to vary these RPUConfigs for inference. It should always be RPUInferenceConfig. What you could vary for inference is the programming noise model and the drift etc. You need to call `model.drift_analog_weights` or model.program_analog_weights` for this. See example https://github.com/IBM/aihwkit/blob/master/examples/05_simple_layer_hardware_aware.py#L91

File: notebooks/examples/analog_sensitivity_LeNet5.ipynb [hunk 4]
```diff
@@ -0,0 +1,1483 @@
...
+       "HBox(children=(SelectionSlider(continuous_update=False, description='ADC steps:', index=5, options=(2, 4, 8, 1…"
+      ]
+     },
+     "metadata": {},
+     "output_type": "display_data"
+    }
+   ],
+   "source": [
+    "devices = widgets.Dropdown(\n",
+    "    options=['FloatingPointDevice', 'ConstantStepDevice', 'LinearStepDevice', 'IdealDevice', 'PulsedDevice', 'UnitCell', 'DigitalRankUpdateCell'],\n",
```
Comment (maljoras) [inline notebooks/examples/analog_sensitivity_LeNet5.ipynb:671]: better to vary the noise model parameters or the t inference time

Comment (maljoras) [inline notebooks/examples/analog_sensitivity_LeNet5.ipynb:671]: better to vary the noise model parameters or the t inference time

File: notebooks/examples/analog_sensitivity_LeNet5.ipynb [hunk 5]
```diff
@@ -0,0 +1,1483 @@
...
+   "outputs": [
+    {
+     "data": {
+      "application/vnd.jupyter.widget-view+json": {
+       "model_id": "6769a08682f1416e83cd2eb7b209f02f",
+       "version_major": 2,
+       "version_minor": 0
+      },
+      "text/plain": [
+       "Dropdown(description='Device:', options=('FloatingPointDevice', 'ConstantStepDevice', 'LinearStepDevice', 'Ide…"
```
Comment (maljoras) [inline notebooks/examples/analog_sensitivity_LeNet5.ipynb:771]: RPUConfig should not be varied for inference chips. Only InferenceRPUConfig

File: notebooks/examples/analog_sensitivity_LeNet5.ipynb [hunk 6]
```diff
@@ -0,0 +1,1483 @@
...
...XvrSly61pvuy+PPOO6/ccMMNM3/vJvPuu+9eXv7yl5c11lhjhbZlcH8BgVZ/w0m9g2Amk+WWuQm0crdprnTRnXWPW+YmlMnd2OV21iu7XCCrNOfqbgKtulELI1ZqoPU//+f/LN2vFi55bbzxxuWP//iPy5Zbbjn6YveHH3549PHDr3/96+Waa64pDzzwwGjoqquuWj74wQ+W7tcNl/X6zne+U84+++xHhWUbbLDB6Evjd9xxxxZ6NzXHINAabqsEM1lvuGVuAq3cbZorXXRn3eOWuQllcjd2uZ31yi4XyCrNubqbQKtu1MKIlRZodU9SvelNb5ox7IKs7qOBXVi1tNePfvSj8u53v7v84he/GP35D//wD8thhx223B489NBD5corryx/8zd/U7qnwZa8ulBrjz32aKF/U3EMAq3htkkwk/WGW+Ym0MrdprnSRXfWPW6Zm1Amd2OX21mv7HKBrNKcq7sJtOpGLYxYaYHWddddV0499dSRYbcgTz755GWGWUugv/SlL5XTTz999J/d01zdE16zeXVh1l//9V+Xq6++evTU1z777FNe8YpXzKbUmDEICLTGgDihtxDMZLDcMjeBVu42zZUuurPuccvchDK5G7vcznpllwtkleZc3U2gVTdqYcRKC7Quu+yycuGFF44Md9hhh/K2t72t6vnv//7v5YgjjhiNW2211cr5559frXnkgFtvvXX0McTnPOc5Aq0Vkus3WKDVz2+S1YKZTJdb5ibQyt2mudJFd9Y9bpmbUCZ3Y5fbWa/scoGs0pyruwm06kYtjFhpgdYVV1wxCpe6VxcwHXPMMVXP73//++XYY48djXvc4x5XzjrrrGrN0gZ0H3fsvk/La24EBFpz45xsRTCTqJXCLXMTaOVu01zpojvrHrfMTSiTu7HL7axXdrlAVmnO1d0EWnWjFkastEDrn//5n8t73/vekWE32boved9ss82Wa9p9RLH7qGL3evrTn17e8573tNCD5o9BoDXcFgtmst5wy9wEWrnbNFe66M66xy1zE8rkbuxyO+uVXS6QVZpzdTeBVt2ohRErLdBavHhxefOb3zzzJe/rrbdeOfTQQ8sLXvCCx7guWrSoXHTRReWLX/zizN+6L5D/kz/5kxZ60PwxCLSG22LBTNYbbpmbQCt3m+ZKF91Z97hlbkKZ3I1dbme9sssFskpzru4m0Kob
```
Comment (maljoras) [inline notebooks/examples/analog_sensitivity_LeNet5.ipynb:960]: Best to mention here that retraining is done in software using the non-idealities in the forward pass but perfect backward pass and update. You can also link to the explanation in the docs: https://aihwkit.readthedocs.io/en/latest/hwa_training.html

Comment (maljoras) [inline notebooks/examples/analog_sensitivity_LeNet5.ipynb:1177]: Here these different device models make sense. One could even run the different optimizers. We actually have a number of presets which you could vary here in the drop down menu, see https://github.com/IBM/aihwkit/blob/master/src/aihwkit/simulator/presets/configs.py

Comment (maljoras) [inline notebooks/examples/analog_sensitivity_LeNet5.ipynb:1177]: Here these different device models make sense. One could even run the different optimizers. We actually have a number of presets which you could vary here in the drop down menu, see https://github.com/IBM/aihwkit/blob/master/src/aihwkit/simulator/presets/configs.py

Comment (maljoras) [inline notebooks/examples/analog_sensitivity_LeNet5.ipynb:1323]: There are also many more update related settings, for instance:
* `rpu_config.backward`  noise  + adc should be set as well (you are only setting the forward pass adc etc)
* `rpu_config.update` has a number of settings defining the update behavior like setting the number of desired pulses or whether to use stochastic update or deterministic bit lines etc.
* `rpu_config.device` settings are now becoming important, defining the response curve and variability of the devices etc.    

Not sure whether it would be too much to look into this in detail but maybe one can select one or two items to vary with the drop down menu that are more targeted for the analog in-memory training

Comment (maljoras) [inline notebooks/examples/analog_sensitivity_LeNet5.ipynb:1323]: There are also many more update related settings, for instance:
* `rpu_config.backward`  noise  + adc should be set as well (you are only setting the forward pass adc etc)
* `rpu_config.update` has a number of settings defining the update behavior like setting the number of desired pulses or whether to use stochastic update or deterministic bit lines etc.
* rpu_config.device settings are now becoming important, defining the response curve and variability of the devices etc.

**Comment by maljoras** [comment]
Hi @AbinandN:
did you have time to look into the comments I made above?

**Comment by AbinandN** [comment]
Yes, @maljoras! Thanks for the very sound suggestions. 

I apologize for not getting back to you sooner. I was caught up with some deadlines for a conference submission. I'll finish up the suggested changes in a day or two.

**Comment by maljoras** [comment]
@AbinandN No problem! I did not want to press you,  just wanted to ask whether you have seen the review. Thanks!

**Comment by AbinandN** [comment]
I have implemented the suggested changes, @maljoras. I feel experimenting with any more parameters in this notebook will make it a bit crowded. Please let me know if you feel otherwise; I can add exploration of a few more settings. Thanks!

**Comment by Fabio-83** [comment]
@AbinandN thanks for your great work. 
I just reorg a bit the introduction to highlight also the strenght of the analog comuting. Could you update the notebook with this code snippet?

```
# Experiments with `aihwkit`
## Exploring the sensitivity of analog inference, analog-aware retraining and in-analog training to the non-idealities of Analog Crossbars

In-memory computing uses crossbar arrays of non-volatile memories to perform Vector-Matrix Multiplications (VMMs) in-situ, i.e., in the memory where the matrix is stored, reducing the need of data movement and eliminating the von Neumann bottleneck. 
The VMM operation is one of the fundamental operations in the workload of neural networks, where vectors of activations are multiplied with arrays of weights. In analog crossbar implementations, the arrays of weights are programmed into memories and the vectors of activations are presented along the wordlines after being converted into analog values by arrays of DACs.
As such, the computation is performed in analog fashion and requires peripheral cicuitry to interface with other digital elements. Once the multiplication is carried out in the analog domain in a fully parallel fashion, the sum-of-products produced on the bitlines are converted into digital values using ADCs for further propagation down the network. The figure below shows the general architecture of a crossbar array along with zoomed-in illustrations of the resistive memory cell and ADC circuitry.

| ![Crossbar architecture](https://github.com/AbinandN/aihwkit/blob/master/notebooks/examples/imgs/xbar.png?raw=1) |
|:--:| 
| **(a) General Architecture of a Crossbar array with illustrations of (b) synaptic device and (c) ADC circuitry** (Reference: https://www.researchgate.net/publication/328091629_Memristor-based_Deep_Convolution_Neural_Network_A_Case_Study) |
 
The strenght of the in-memory analog computing lies in the use of the non-volatile memory crossbar architecture to fully parallelize VMM (with a boost in operations per second) while decreasing the need of data movement (giving a siginifant reduction of the energy needed to run the array). 
However, analog computations introduce non-idealitites and tradeoffs that could impact the accuracy of neural networks.

In an crossbar array, the weights are stored in synaptic devices called Resistive Processing Units (RPUs), which model analog memory cells. A two-dimensional array of RPUs makes up an Analog Tile in the toolkit.

This notebook studies the sensitivity of a crossbar implementation of a LeNet5 network trained on the MNIST dataset for handwritten digit recognition to a range of non-idealities introduced by the peripheral circuits and choice of synaptic device.
```

**Comment by AbinandN** [comment]
Thanks, @Fabio-83 !

I have changed the intro as suggested.

**Review by maljoras — APPROVED** [review]
Thank you very much! This is a great addition
